rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,473 @@
+      SE_Device* executor, SE_Stream stream,
+      SE_DeviceMemoryBase* device_dst,
+      const void* host_src, uint64_t size);
+
+  // Causes the host code to synchronously wait for operations entrained onto
+  // stream to complete. Effectively a join on the asynchronous device
+  // operations enqueued on the stream before this program point.
+  void (*block_host_until_done)(
+      SE_Device* executor, SE_Stream stream,
+      TF_Status* status);
+
+  // Synchronizes all activity occurring in the StreamExecutor's context (most
+  // likely a whole device).
+  bool (*synchronize_all_activity)(SE_Device* executor);
+
+  // Obtains metadata about the underlying device.
+  void fill_device_description(SE_Device* executor,
+                               SE_DeviceDescription* description,
+                               TF_Status* status);
+
+  // Entrains on a stream a user-specified function to be run on the host.
+  bool host_callback(SE_Device* executor, SE_Stream* stream,
+                     TF_StatusCallbackFn callback_fn, void* ctx);
+} SE_Device;
+
+TF_CAPI_EXPORT extern SE_Platform* SE_NewPlatform(
+     SE_PlatformId id,
+     int visible_device_count,
+     SE_Device* (*create_device)(SE_Options* options, TF_Status* status),
+     SE_StreamExecutor* (*create_stream_executor)(TF_Status* status),
+     void (*delete_device)(SE_Device* device),
+     void (*delete_stream_executor)(SE_StreamExecutor* stream_executor);
+);
+
+TF_CAPI_EXPORT extern void SE_RegisterPlatform(
+     char* name,
+     size_t name_len,
+     SE_Platform* platform,
+     TF_Status* status);
+
+#ifdef __cplusplus
+} /* end extern "C" */
+#endif
+```
+
+## PlatformId
+
+`SE_PlatformId.id` should be set to a unique identifier.
+
+## Usage example
+
+Define functions that create and delete `SE_Device` and `SE_StreamExecutor`:
+
+```cpp
+SE_Device* create_device(SE_Options* options, TF_Status* status) {
+  SE_Device* se = new SE_Device();
+  se->device_handle = get_my_device_handle();
+  ...
+  return se;
+}
+SE_StreamExecutor* create_stream_executor(TF_Status* status) {
+  SE_StreamExecutor* se_fns = new SE_StreamExecutor();
+  se->memcpy_from_host = my_device_memcpy_from_host_function;
+  ...
+  return se;
+}
+void delete_device(SE_Device* device) {
+  -- delete device handle here --
+  ...
+  delete device;
+}
+void delete_stream_executor(SE_StreamExecutor* stream_executor) {
+  delete stream_executor;
+}
+```
+
+Create a new platform using `SE_NewPlatform` and register it using
+`SE_RegisterPlatform`:
+
+```cpp
+void RegisterMyCustomPlatform() {
+  static plugin_id_value = 123;
+  SE_PlatformId id;
+  id.id = &plugin_id_value;
+  int visible_device_count = 2;
+
+  SE_Platform* custom_platform = SE_NewPlatform(
+     id, visible_device_count,
+     create_device, create_stream_executor,
+     delete_device, delete_stream_executor);
+
+  TF_Status* status = TF_NewStatus();
+  std::string name = "MyCustomDevice";
+  SE_RegisterPlatform(
```

Comment:
Hi, @annarev  I have  little confuse about this part, what SE_RegisterPlatform did and where it will register itself to, in current StreamExecutor TPU code, it seems that it will have C++ object: TpuPlatform and it will be register to MultiPlatformManager which is the same as other platform(cuda, rocm), and SE_Platform will be a  member(opaque handle) in TpuPlatform, so I am confuse what SE_RegisterPlatform did here and how it will connect with MultiPlatformManager. Thanks very much.

Reply:
It will be the same here. The call will register a new platfrom with MultiPlatfromManager.

Reply:
Ok, Thanks